// diva-shared-apps/components/multiselect/multiselect.component.ts
/**
 * This file should contains all the logic and parameters related to multiselect dropdown.
 * Any changes in this file will lead to changes in all the places
 * where ever its refered
 */
import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { SelectItem } from 'primeng/api';

@Component({
  selector: 'app-multiselect',
  templateUrl: './multiselect.component.html',
  styleUrls: ['./multiselect.component.css']
})
export class MultiselectComponent implements OnInit {


  @Input() selectedItems: string[];
  @Input() defaultlabel: string;
  @Output() selectedMultiSelect = new EventEmitter();
  display: boolean;
  filtereddatasource: any[] = [];
  originaldatasource: SelectItem[];
  @Input() size: number;
  @Input() disabled: boolean;
  @Input() multiselectheader;
  @Input()
  set dataSource (data: SelectItem[]) {
    console.log('data', data);
    this.filtereddatasource = [];
    this.originaldatasource = data;
    if (data && data.length > this.size) {
      for (let i = 0; i <= this.size; i++) {
        this.filtereddatasource.push(data[i]);
      }
    } else if (data) {
      this.filtereddatasource = data;
    }
    console.log('filtered Data', this.filtereddatasource);
  }

  get dataSource () {
    return this.filtereddatasource;
  }
  constructor() {
    this.defaultlabel = 'Select option(s)';
    this.size = 500;
  }

  ngOnInit() {
  }
  /* when ever the change event is triggered, check whether value has been selected.
  Incase value has been deselected then the value would be null. Incase of null then trigger the event with none*/
  filterdropdown(event) {
    console.log('filterdropdown', event);
    this.filtereddatasource = [];
    if (event) {
      for (const item of this.originaldatasource) {
        if (item.label.toLowerCase().indexOf(event.toLowerCase()) === 0 && this.filtereddatasource.length < this.size) {
          this.filtereddatasource.push(item);
        } else if (this.filtereddatasource.length === this.size) {
          break;

        }
      }
    } else if (this.originaldatasource.length > this.size) {
      for (let i = 0; i <= this.size; i++) {
        this.filtereddatasource.push(this.originaldatasource[i]);
      }
    } else {
      this.filtereddatasource =  this.originaldatasource;
    }
    console.log('filteredcontrols', this.filtereddatasource);
  }
  change(event) {
    if (event.value && event.value.length > 0) {
      this.selectedMultiSelect.emit(event.value);
    } else {
      this.selectedMultiSelect.emit('none');
    }

  }

  displayModel() {
    this.display = true;
  }
}
